crates/red_knot_python_semantic/src/types.rs

```diff
@@ -235,23 +237,56 @@ impl<'db> Type<'db> {
             }
             Type::IntLiteral(_) => {
                 // TODO raise error
-                Type::Unknown
+                Type::Unknown(UnknownTypeKind::RedKnotLimitation)
             }
-            Type::BooleanLiteral(_) => Type::Unknown,
+            Type::BooleanLiteral(_) => Type::Unknown(UnknownTypeKind::RedKnotLimitation),
         }
     }
 
     #[must_use]
     pub fn instance(&self) -> Type<'db> {
         match self {
             Type::Any => Type::Any,
-            Type::Unknown => Type::Unknown,
+            Type::Unknown(kind) => Type::Unknown(*kind),
             Type::Class(class) => Type::Instance(*class),
-            _ => Type::Unknown, // TODO type errors
+            _ => Type::Unknown(UnknownTypeKind::RedKnotLimitation), // TODO type errors
         }
     }
 }
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum UnknownTypeKind {
+    /// Temporary variant that indicates that we *should*
+    /// be able to infer a type here in due course, but currently can't
+    RedKnotLimitation,
+
+    /// Invalid syntax in the node means we can't infer a type here
+    InvalidSyntax,
+
+    /// An `Unknown` type stemming from some kind of type error.
+    ///
+    /// Examples:
+    /// - A function was called with argument(s) of incorrect type(s), leading
+    ///   to a diagnostic being emitted and the expression being evaluated as `Unknown`.
+    /// - An expression was deemed to take place between two invalid types, e.g. `1 + "foo"`.
+    ///   (This is really the same as the first example, since expressions ultimately desguar
+    ///   to function calls, e.g. `1 + "foo"` desugars approximately to `type(1).__add__("foo")`.)
+    /// - An attribute or method was looked up on an instance of a type that doesn't have that
+    ///   attribute or method.
+    TypeError,
```

Comment:
This covers quite a lot. If the intention is to only ever emit diagnostics after all types have been inferred, we may need to split this into several variants, so that we can emit different error messages (under different codes) for different kinds of errors. On the other hand, we could consider collecting some diagnostics as part of the `TypeInferenceBuilder`, which would mean we wouldn't need so much granularity here.